lib/Basic/Unix/TaskQueue.inc

```diff
@@ -320,164 +321,320 @@ void TaskQueue::addTask(const char *ExecPath, ArrayRef<const char *> Args,
   QueuedTasks.push(std::move(T));
 }
 
-bool TaskQueue::execute(TaskBeganCallback Began, TaskFinishedCallback Finished,
-                        TaskSignalledCallback Signalled) {
+// Owns Tasks, handles correspondence with FDs and pids.
+// FIXME: only handles stdout fds.
+class TaskMap {
+  using PidToTaskMap = llvm::DenseMap<pid_t, std::unique_ptr<Task>>;
+  PidToTaskMap TasksByPid;
+
+public:
+  TaskMap() = default;
+
+  bool empty() const { return TasksByPid.empty(); }
+  unsigned size() const { return TasksByPid.size(); }
+
+  void add(std::unique_ptr<Task> T) { TasksByPid[T->getPid()] = std::move(T); }
+
+  Task &findTaskForFd(const int fd) {
+    auto predicate = [&fd](PidToTaskMap::value_type &value) -> bool {
+      return value.second->getPipe() == fd;
+    };
+    auto iter = std::find_if(TasksByPid.begin(), TasksByPid.end(), predicate);
+    assert(iter != TasksByPid.end() &&
+           "All outstanding fds must be associated with a  Task");
+    return *iter->second;
+  }
+
+  void eraseTask(Task &T) { TasksByPid.erase(T.getPid()); }
+};
+
+/// Concurrently execute the tasks in the TaskQueue, collecting the outputs from
+/// each task. The typical task is a Swift frontend job.
```

Comment:
> "The typical task is a Swift frontend job."

It feels a little weird to put this here. In theory this TaskQueue class could be sunk down to LLVM as a general utility. There's (deliberately) nothing Swift-specific about it.

Reply:
I agree. This class basically replaces a function that was here before. I broke up the function into pieces, members of the new class. The function could have been moved to LLVM, I guess.